### ISSUE TITLE
Error 2345 when options are a variable (but not if options are provided inline)

### ISSUE THREAD

**Issue by Steve-Mcl** [microsoft/TypeScript#44613]
# Bug Report

### 🔎 Search Terms
2345 and DateTimeFormat

I found a lot of issues similar but nothing about where a in-line code is OK and a variable is not OK

### 🕗 Version & Regression Information
- This changed between versions 4.1.5 and 4.23

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAJgQygUxgXhmZB3GARJZACgEoBuAKAtElhAAcoBLcCdGAbxm2WQGtEATwBcMAOQAbcAHMxAGhiDkCAE6ixYAK4BbZCqbB5MbeCgALdVLCyFQ9Vt37DMAL6VqrWHAgBGdllwASTAoCQA6AhQAFSZdADEQFW0kIjE4ZABaPABRIwZmVhIwgDNE5KgiRBRyD1p4CAAmfxwYYNCIwhj4spS0zJyjLh5+O3ErG0VlNXEHPQMjExCLMZkjUY0dOecXItKklKrkGoogA)

### 💻 Code

```ts
const date = new Date();

const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };

const ds1 = new Intl.DateTimeFormat('de-DE', options).format(date);
const ds2 = new Intl.DateTimeFormat('de-DE', { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' }).format(date);

```

### 🙁 Actual behaviour
options object (`Intl.DateTimeFormat` options 2nd parameter)  is flagged as an error when provided as a variable but not when they are provided in-line...

![image](https://user-images.githubusercontent.com/44235289/122249858-1a4a0280-cec1-11eb-9ff5-d5f0e6c67033.png)

### 🙂 Expected behaviour

No error as per V4.1.5 and below
![image](https://user-images.githubusercontent.com/44235289/122249918-2766f180-cec1-11eb-82eb-db9cb1509938.png)

### ℹ️ Info
Originally posted here issue in monaco-editor here: https://github.com/microsoft/monaco-editor/issues/2531 but the issue was closed and I was told to raise it here.

**Comment by RyanCavanaugh** [comment]
Specifically, you should be writing
```ts
const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' } as const;
```

**Comment by Steve-Mcl** [comment]
@MartinJohns @RyanCavanaugh 

Thank you very much for responding. I still however have a dilemma - I hope you can help or point me in the right direction,

Originally this issue was raised in monaco-editor because the code in question is `JavaScript`. That means we cannot use `const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' } as const;`


![image](https://user-images.githubusercontent.com/44235289/122259104-b1b35380-cec9-11eb-9551-29c76381768c.png)


See microsoft/monaco-editor#2531 and test in [monaco playground](https://microsoft.github.io/monaco-editor/playground.html)  (code below)


<details>
  <summary>Click to see monaco playground code</summary>

```js

//PASTE THIS INTO LEFT-HAND EDITOR THEN CLICK RUN

const defCode = `
const date = new Date();

const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };

const ds1 = new Intl.DateTimeFormat('de-DE', options).format(date);
const ds2 = new Intl.DateTimeFormat('de-DE', { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' }).format(date);
`
const compilerOptions = {
    allowJs: true,
    checkJs: true,
    allowNonTsExtensions: true,
    target: monaco.languages.typescript.ScriptTarget.ESNext,
    strictNullChecks: false,
    strictPropertyInitialization: true,
    strictFunctionTypes: true,
    strictBindCallApply: true,
    moduleResolution: monaco.languages.typescript.ModuleResolutionKind.NodeJs,
    module: monaco.languages.typescript.ModuleKind.CommonJS,
    lib: ["esnext"] //dont load DOM by default,
}
monaco.languages.typescript.javascriptDefaults.setCompilerOptions(compilerOptions);

const diagnosticOptions = {
    noSemanticValidation: false,
    noSyntaxValidation: false,
    diagnosticCodesToIgnore:  [
        1108,  //return not inside function
        //2304,  //Cannot find name - this one is heavy handed and prevents user seeing stupid errors. Would provide better ACE feature parity (i.e. no need for declaration of vars) but misses lots of errors. Lets be bold & leave it out!
        2307,  //Cannot find module 'xxx' or its corresponding type declarations
        2322,  //Type 'unknown' is not assignable to type 'string'
        2339,  //property does not exist on
        //2345,  //Argument of type xxx is not assignable to parameter of type 'DateTimeFormatOptions' 
        7043,  //i forget what this one is,
        80001, //Convert to ES6 module
        80004, //JSDoc types may be moved to TypeScript types.
    ]
};
monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions(diagnosticOptions);

window.editor = monaco.editor.create(document.getElementById("container"), {
value: defCode,
language: "javascript"
});

```

</details>


Do either of you know how this can be handled or where this issue should be raised? (monaco-core?, monaco-editor?, here?)?

**Comment by fatcerberus** [comment]
A bit more insight in case you were curious: The problem occurs because `.weekday` is required to be one of `"long" | "short" | "narrow"` but assigning the object literal to a variable first, with no further context, causes it to be widened to `string`, which TS then can't verify is compatible when you pass it to the function later.  Passing the object literal directly works because contextual typing prevents the widening (i.e. TS can verify on the spot that you're passing a compatible string literal).

**Comment by fatcerberus** [comment]
As for a workaround, you can try this:
```js
/** @type {Intl.DateTimeFormatOptions} */
const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
```

**Comment by Steve-Mcl** [comment]
Thank you @fatcerberus, very insightful. Not 100% sure the editor shouldn't be able to figure this out considering no changes between object creation and use, but I understand now why.

Cheers.

**Comment by Steve-Mcl** [comment]
@fabiancook, @RyanCavanaugh  can I bother you one guys more time please... 

How come all versions of TypeScript up to (and including) V4.1.5 dont have this problem?

![image](https://user-images.githubusercontent.com/44235289/122668663-12d07500-d1b1-11eb-9ff2-a9edcec53773.png)

**Comment by MartinJohns** [comment]
> How come all versions of TypeScript up to (and including) V4.1.5 dont have this problem?

Because with updated TypeScript versions you get updated library definitions as well.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-2.html#libdts-updates
